README.md
RL_JSON
=======

This package adds a command [json] to the interpreter, and defines a new
Tcl_Obj type to store the parsed JSON document.  The [json] command directly
manipulates values whose string representation is valid JSON, in a similar
way to how the [dict] command directly manipulates values whose string
representaion is a valid dictionary.  It is similar to [dict] in performance.

Also provided is a command [json template] which generates JSON documents
by interpolating values into a template from a supplied dictionary or
variables in the current call frame.  The templates are valid JSON documents
containing string values which match the regex "^~[SNBJTL]:.+$".  The second
character determines what the resulting type of the substituted value will be:
* S: A string.
* N: A number.
* B: A boolean.
* J: A JSON fragment.
* T: A JSON template (substitutions are performed on the inserted fragment).
* L: A literal - the resulting string is simply everything from the forth character onwards (this allows literal strings to be included in the template that would otherwise be interpreted as the substitutions above).

The value inserted is determined by the characters following the substitution
type prefix.  When interpolating values from a dictionary they name keys in the
dictionary which hold the values to interpolate.  When interpolating from
variables in the current scope, they name scalar or array variables which hold
the values to interpolate.  In either case if the named key or variable doesn't
exist, a JSON null is interpolated in its place.

Quick Reference
---------------
* [json get *json_val* ?*key* ... ?*modifier*??]  - Extract the value of a portion of the *json_val*, returns the closest native Tcl type (other than JSON) for the extracted portion.
* [json parse *json_val*]  - A deprecated synonym for [json get *json_val*].
* [json get_typed *json_val* ?*key* ... ?*modifier*??]  - Extract the value of a portion of the *json_val*, returns a two element list: the first being the value that would be returned by [json get] and the second being the JSON type of the extracted portion.
* [json extract *json_val* ?*key* ... ?*modifier*??]  - Extract the value of a portion of the *json_val*, returns the JSON fragment.
* [json exists *json_val* ?*key* ... ?*modifier*??]  - Tests whether the supplied key path and modifier resolve to something that exists in *json_val*
* [json normalize *json_val*]  - Return a "normalized" version of the input *json_val* - all optional whitespace trimmed.
* [json template *json_val* ?*dictionary*?]  - Return a JSON value by interpolating the values from *dictionary* into the template, or from variables in the current scope if *dictionary* is not supplied, in the manner described above.
* [json new *type* *value*]  - Return a JSON fragment of type *type* and value *value*.
* [json fmt *type* *value*]  - A deprecated synonym for [json new *type* *value*].
* [json foreach *varlist1* *json_val1* ?*varlist2* *json_val2* ...? *script*]  - Evaluate *script* in a loop in a similar way to the [foreach] command.  In each iteration, the values stored in the iterator variables in *varlist* are the JSON fragments from *json_val*.  Supports iterating over JSON arrays and JSON objects.  In the JSON object case, *varlist* must be a two element list, with the first specifiying the variable to hold the key and the second the value.  In the JSON array case, the rules are the same as the [foreach] command.
* [json lmap *varlist1* *json_val1* ?*varlist2* *json_val2* ...? *script*]  - As for [json foreach], except that it is collecting - the result from each evaluation of *script* is added to a list and returned as the result of the [json lmap] command.  If the *script* results in a TCL_CONTINUE code, that iteration is skipped and no element is added to the result list.  If it results in TCL_BREAK the iterations are stopped and the results accumulated so far are returned.

Paths
-----

The commands [json get], [json get_typed], [json extract] and [json exists]
accept a path specification that names some subset of the supplied *json_val*.
The rules are similar to the equivalent concept in the [dict] command, except
that the paths used by [json] allow indexing into JSON arrays by the integer
key (or a string matching the regex "^end(-[0-9]+)?$"), and that the last
element can be a modifier:
* **?type** - Returns the type of the named fragment.
* **?length** - When the path refers to an array, this returns the length of the array.  When the path refers to a string, this returns the number of characters in the string.  All other types throw an error.
* **?size** - Valid only for objects, returns the number of keys defined in the object.
* **?keys** - Valid only for objects, returns a list of the keys in the object.

A literal value that would match one of the above modifiers can be used as the last element in the path by doubling the ?:

~~~tcl
json get {
    {
        "foo": {
            "?size": "quite big"
        }
    }
} foo ??size
~~~

Returns "quite big"

Examples
--------

Produce a JSON value from a template:
~~~tcl
json template {
    {
        "thing1": "~S:val1",
        "thing2": ["a", "~N:val2", "~S:val2", "~B:val2", "~S:val3", "~L:~S:val1"],
        "subdoc1": "~J:subdoc",
        "subdoc2": "~T:subdoc"
    }
} {
    val1   hello
    val2   1e6
    subdoc {
        { "thing3": "~S:val1" }
    }
}
~~~
Result:
~~~json
{"thing1":"hello","thing2":["a",1000000.0,"1e6",true,null,"~S:val1"],"subdoc1":{"thing3":"~S:val1"},"subdoc2":{"thing3":"hello"}}
~~~

## Performance

Good performance was a requirement for rl_json, because it is used to handle
large volumes of data flowing to and from various JSON based REST apis.  It's
generally the fastest option for working with JSON values in Tcl from the
options I've tried, with the next closest being yajltcl.  These benchmarks
report the median times in microseconds, and produce quite stable results
between runs.  Benchmarking was done on a MacBook Air running Ubuntu
14.04 64bit, Tcl 8.6.3 built with -O3 optimization turned on, and using
an Intel i5 3427U CPU.

### Parsing

This benchmark compares the relative performance of extracting the field
containing the string "obj" from the JSON doc:

```json
{
	"foo": "bar",
	"baz": ["str", 123, 123.4, true, false, null, {"inner": "obj"}]
}
```

The compared methods are:

| Name | Notes | Code |
|------|-------|------|
| old_json_parse | Pure Tcl parser | `dict get [lindex [dict get [json_old parse [string trim $json]] baz] end] inner` |
| rl_json_parse | | `dict get [lindex [dict get [json parse [string trim $json]] baz] end] inner` |
| rl_json_get | Using the built-in accessor method | `json get [string trim $json] baz end inner` |
| yajltcl | | `dict get [lindex [dict get [yajl::json2dict [string trim $json]] baz] end] inner` |
| rl_json_get_native | | `json get $json baz end inner` |

The use of [string trim $json] is to defeat the caching of the parsed
representation, forcing it to reparse the string each time since we're
measuring the parse performance here.  The exception is the rl_json_get_native
test which demonstrates the performance of the cached case.

```
-- parse-1.1: "Parse a small JSON doc and extract a field" --------------------
                   | This run
    old_json_parse |  241.595
     rl_json_parse |    5.540
       rl_json_get |    4.950
           yajltcl |    8.800
rl_json_get_native |    0.800
```

### Generating

This benchmark compares the relative performance of various ways of
dynamically generating a JSON document.  Although all the methods produce the
same string, only the "template" and "template_dict" variants handle nulls in
the general case - the others manually test for null only for the one field
that is known to be null, so the performance of these variants would be worse
in a real-world scenario where all fields would need to be tested for null.

The JSON doc generated in each case is the one produced by the following JSON
template (where a(not_defined) does not exist and results in a null value in the produced document):

```json
{
	"foo": "~S:bar",
	"baz": [
		"~S:a(x)",
		"~N:a(y)",
		123.4,
		"~B:a(on)",
		"~B:a(off)",
		"~S:a(not_defined)",
		"~L:~S:not a subst",
		"~T:a(subdoc)",
		"~T:a(subdoc2)"
	]
}
```

The produced JSON doc is:

```json
{"foo":"Bar","baz":["str\"foo\nbar",123,123.4,true,false,null,"~S:not a subst",{"inner":"Bar"},{"inner2":"Bar"}]}
```

The code for these variants are too long to include in this table, refer to
bench/new.bench for the details.

| Name | Notes |
|------|-------|
| old_json_fmt | Pure Tcl implementation, builds JSON from type-annotated Tcl values |
| rl_json_new | rl_json's [json new], API compatible with the pure Tcl version used in old_json_fmt |
| template | rl_json's [json template] |
| yajltcl | yajltcl's type-annotated Tcl value approach |
| template_dict | As for template, but using a dict containing the values to substitute |
| yajltcl_dict | As for yajltcl, but extracting the values from the same dict used by template_dict |

```
-- new-1.1: "Various ways of dynamically assembling a JSON doc" ---------------
                 | This run
    old_json_fmt |   49.450
     rl_json_new |   10.240
        template |    4.520
         yajltcl |    7.700
   template_dict |    2.500
    yajltcl_dict |    7.530
```

Under the Hood
--------------

The yajl c library is used to parse the JSON string, and to generate properly
quoted strings when serializing JSON values.  JSON values are parsed to an
internal format using Tcl_Objs and stored as the internal representation for
a new type of Tcl_Obj.  Subsequent manipulation of that value use the internal
representation directly.

License
-------

Copyright 2015 Ruby Lane.  Licensed under the same terms as the Tcl core.


tclconfig/ChangeLog
2007-05-18  Donal K. Fellows  <donal.k.fellows@man.ac.uk>

	* tcl.m4: Added quoting so that paths with spaces cause fewer
	problems.

2007-03-07  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): s/CFLAGS/CPPFLAGS/ in -mmacosx-version-min check.

2007-02-15  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: correct private header check to search in generic subdir

2007-02-09  Jeff Hobbs  <jeffh@ActiveState.com>

	*** Bump to TEA_VERSION 3.6 ***

	* tcl.m4: correct -d to -f
	(TEA_CONFIG_CFLAGS): SHLIB_SUFFIX is .so on HP ia64 [Bug 1615058]
	
2007-02-08  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4 (TEA_PRIVATE_TCL_HEADERS, TEA_PRIVATE_TK_HEADERS): check
	that the dirs actually have private headers. [Bug 1631922]

2007-02-04  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4: add caching to -pipe check.

2007-01-25  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4: integrate CPPFLAGS into CFLAGS as late as possible and 
	move (rather than duplicate) -isysroot flags from CFLAGS to CPPFLAGS to
	avoid errors about multiple -isysroot flags from some older gcc builds.

2006-01-19  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4: ensure CPPFLAGS env var is used when set. [Bug 1586861]
	(Darwin): add -isysroot and -mmacosx-version-min flags to CPPFLAGS when
	present in CFLAGS to avoid discrepancies between what headers configure
	sees during preprocessing tests and compiling tests.

2006-12-19  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): --enable-64bit: verify linking with 64bit -arch flag
	succeeds before enabling 64bit build.

2006-12-16  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Linux): fix previous change to use makefile variable
	LDFLAGS_DEFAULT instead of LDFLAGS in SHLIB_LD, to ensure linker
	flags in sampleextension Makefile are picked up.

2006-11-26  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Linux): --enable-64bit support. [Patch 1597389], [Bug 1230558]

2006-08-18  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): add support for --enable-64bit on x86_64, for
	universal builds including x86_64 and for use of -mmacosx-version-min
	instead of MACOSX_DEPLOYMENT_TARGET. For Tk extensions, remove 64-bit
	arch flags from CFLAGS like in the Tk configure, as neither TkAqua nor
	TkX11 can be built for 64-bit at present.

2006-03-28  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: []-quote AC_DEFUN functions.
	(TEA_PATH_TKCONFIG): Fixed Windows-specific check for tkConfig.sh.
	(TEA_MAKE_LIB): Prepend 'lib' for Windows-gcc configs.

2006-03-07  Joe English  <jenglish@users.sourceforge.net>

	* tcl.m4: Set SHLIB_LD_FLAGS='${LIBS}' on NetBSD,
	as per the other *BSD variants [Bug 1334613].

2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	*** Bump to TEA version 3.5 ***

	* tcl.m4: keep LD_SEARCH_FLAGS and CC_SEARCH_FLAGS synchronous
	with core tcl.m4 meaning.

2006-01-24  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): use makefile variable LDFLAGS_DEFAULT instead of
	LDFLAGS in SHLIB_LD, to ensure linker flags in sampleextension Makefile
	are picked up. [Bug 1403343]

2006-01-23  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: add C:/Tcl/lib and C:/Progra~1/Tcl/lib dirs to check for
	*Config.sh on Windows. [Bug 1407544]

2006-01-23  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): for Tk extensions, remove -arch ppc64 from CFLAGS
	like in the Tk configure, as neither TkAqua nor TkX11 can be built for
	64bit at present (no 64bit GUI libraries).

2006-01-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: restore system=windows on Windows.
	Remove error if 'ar' isn't found (it may not be on Windows).
	Do not add -lxnet or define _XOPEN_SOURCE on HP-UX by default.
	Ensure the C|LDFLAGS_DEFAULT gets the fully sub'd value at
	configure time.

2006-01-10  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4: add caching, use AC_CACHE_CHECK instead of AC_CACHE_VAL
	where possible, consistent message quoting, sync relevant
	tcl/unix/tcl.m4 HEAD changes and gratuitous formatting differences
	(notably sunc removal of support for for ancient BSD's, IRIX 4,
	RISCos and Ultrix by kennykb), Darwin improvements to
	TEA_LOAD_*CONFIG to make linking work against Tcl/Tk frameworks
	installed in arbitrary location, change TEA_PROG_* search order
	(look in *_BIN_DIR parents before *_PREFIX).

2006-01-05  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: add dkf's system config refactor

2006-01-04  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: remove extraneous ' that causes bash 3.1 to choke

2005-12-19  Joe English  <jenglish@users.sourceforge.net>

	* tcl.m4 (TEA_PATH_TCLCONFIG &c): Look for tclConfig.sh &c 
	in ${libdir}, where they are installed by default [Patch #1377407].

2005-12-05  Don Porter  <dgp@users.sf.net>

	* tcl.m4 (TEA_PUBLIC_*_HEADERS):	Better support for finding
	header files for uninstalled Tcl and Tk.

2005-12-02  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4: correctly bump TEA_VERSION var to 3.4

2005-12-01  Daniel Steffen  <das@users.sourceforge.net>

	* unix/tcl.m4 (Darwin): fixed error when MACOSX_DEPLOYMENT_TARGET unset

2005-11-29  Jeff Hobbs  <jeffh@ActiveState.com>

	* tcl.m4:  *** Bump to TEA version 3.4 ***
	Add Windows x64 build support.
	Remove TEA_PATH_NOSPACE and handle the problem with ""s where
	necessary - the macro relied on TCLSH_PROG which didn't work for
	cross-compiles.

2005-11-27  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4 (Darwin): add 64bit support, add CFLAGS to SHLIB_LD to
	support passing -isysroot in env(CFLAGS) to configure (flag can't
	be present twice, so can't be in both CFLAGS and LDFLAGS during
	configure), don't use -prebind when deploying on 10.4.
	(TEA_ENABLE_LANGINFO, TEA_TIME_HANDLER): add/fix caching.

2005-10-30  Daniel Steffen  <das@users.sourceforge.net>

	* tcl.m4: fixed two tests for TEA_WINDOWINGSYSTEM = "aqua" that
	should have been for `uname -s` = "Darwin" instead; added some
	missing quoting.

perf/all.tcl
#!/usr/bin/env cfkit8.6
# vim: ft=tcl foldmethod=marker foldmarker=<<<,>>> ts=4 shiftwidth=4

if {[file system [info script]] eq "native"} {
	package require platform

	foreach platform [platform::patterns [platform::identify]] {
		set tm_path		[file join $env(HOME) .tbuild repo tm $platform]
		set pkg_path	[file join $env(HOME) .tbuild repo pkg $platform]
		if {[file exists $tm_path]} {
			tcl::tm::path add $tm_path
		}
		if {[file exists $pkg_path]} {
			lappend auto_path $pkg_path
		}
	}
}

set here	[file dirname [file normalize [info script]]]
set parent	[file dirname $here]
tcl::tm::path add $here
lappend auto_path $parent
lappend auto_path /home/cyan/local/lib

package require rl_json
package require benchmark
package require yajltcl
package require json_old

set json {
	{
		"foo": "bar",
		"baz": ["str", 123, 123.4, true, false, null, {"inner": "obj"}]
	}
}

set copy	"$json "
json parse $copy

set tmpl {
	{
		"foo": "~S:bar",
		"baz":  [
			"~S:a(x)",
			"~N:a(y)",
			123.4,
			"~B:a(on)",
			"~B:a(off)",
			"~S:a(not_defined)",
			"~L:~S:not a subst",
			"~J:a(subdoc)",
			"~T:a(subdoc2)"
		]
	}
}

set simple_tmpl {
	{
		"foo": "~S:bar",
		"baz": ["str", 123, 123.4, true, "~B:a(on)", null, {"inner": "obj"}]
	}
}

proc our_parse2dict_forced {} {
	global json
	dict get [lindex [dict get [json parse [string trim $json]] baz] end] inner
}

proc yajltcl_parse2dict_forced {} {
	global json
	dict get [lindex [dict get [yajl::json2dict [string trim $json]] baz] end] inner
}

proc our_parse2dict {} {
	global json
	dict get [lindex [dict get [json parse $json] baz] end] inner
}

proc yajltcl_parse2dict {} {
	global json
	dict get [lindex [dict get [yajl::json2dict $json] baz] end] inner
}


proc our_serialize {} {
	global json

	set copy2	$json
	string cat [json normalize $copy2]
}

proc our_normalize {} {
	global copy
	string cat [json normalize "$copy "]
}

array set a {
	x		str
	y		123
	on		yes
	off		0
	subdoc	{{"inner": "~S:bar"}}
	subdoc2	{{"inner2": "~S:bar"}}
}

proc our_simple_template_vars {} {
	global simple_tmpl a

	set bar	"Bar"

	json from_template $simple_tmpl
}

proc our_simple_template_dict {} {
	global simple_tmpl

	json from_template $simple_tmpl {
		foo		"Foo"
		bar		"Bar"
		baz		"Baz"
		a(x)		str
		a(y)		123
		a(on)		yes
		a(off)		0
		a(subdoc)	{{"inner" : "~S:bar"}}
		a(subdoc2)	{{"inner2" : "~S:bar"}}
	}
}


proc our_template_vars {} {
	global tmpl a

	set bar	"Bar"

	json from_template $tmpl
}

proc our_template_dict {} {
	global tmpl

	json from_template $tmpl {
		foo		"Foo"
		bar		"Bar"
		baz		"Baz"
		a(x)		str
		a(y)		123
		a(on)		yes
		a(off)		0
		a(subdoc)	{{"inner" : "~S:bar"}}
		a(subdoc2)	{{"inner2" : "~S:bar"}}
	}
}


proc naive_template {} {
	global a

	set bar	"Bar"

	subst {
		{
			"foo": "$bar",
			"baz":  \[
				"$a(x)",
				$a(y),
				123.4,
				[expr {$a(on) ? "true" : "false"}],
				[expr {$a(off) ? "true" : "false"}],
				[expr {[info exists a(not_defined)] ? "\"$a(not_defined)\"" : "null"}],
				"~S:not a subst",
				{"inner" : "$bar"},
				{"inner2" : "$bar"}
			\]
		}
	}
}

proc old_template {} {
	global a

	set bar	"Bar"

	json_old fmt obj [list \
		"foo" [list s $bar] \
		"baz" [list a \
			[list s $a(x)] \
			[list num $a(y)] \
			[list num 123.4] \
			[list bool $a(on)] \
			[list bool $a(off)] \
			[expr {[info exists a(not_defined)] ? [list s $a(not_defined)] : [list null]}] \
			[list s "~S:not a subst"] \
			[list o inner [list s $bar]] \
			[list o inner2 [list s $bar]] \
		] \
	]
}

proc new_fmt {} {
	global a

	set bar	"Bar"

	json fmt obj [list \
		"foo" [list s $bar] \
		"baz" [list a \
			[list s $a(x)] \
			[list num $a(y)] \
			[list num 123.4] \
			[list num 0x20] \
			[list bool $a(on)] \
			[list bool $a(off)] \
			[expr {[info exists a(not_defined)] ? [list s $a(not_defined)] : [list null]}] \
			[list s "~S:not a subst"] \
			[list o inner [list s $bar]] \
			[list o inner2 [list s $bar]] \
		] \
	]
}

proc our_serialize_template {} {
	global tmpl

	json normalize $tmpl
}


proc nop {} {
	global json
	json nop
}


proc main {} {
	global json tmpl simple_tmpl a

	puts "rl_json Parsed: [json parse [string trim $json]]"
	puts "yajltcl Parsed: [yajl::json2dict [string trim $json]]"
	puts "serialized: [our_serialize]"
	puts "new fmt: [new_fmt]"

	set bar	"Bar"

	benchmark::test 300 100 usec our_parse2dict_forced
	benchmark::test 300 100 usec yajltcl_parse2dict_forced
	benchmark::test 600 200 usec our_parse2dict
	benchmark::test 300 100 usec yajltcl_parse2dict

	benchmark::test 600 200 usec our_serialize
	benchmark::test 600 200 usec our_normalize
	benchmark::test 600 200 usec json normalize $json
	benchmark::test 600 200 usec json normalize "$json "

	benchmark::test 600 200 usec our_simple_template_vars
	benchmark::test 600 200 usec our_simple_template_dict
	benchmark::test 600 200 usec json from_template $simple_tmpl {
		foo		"Foo"
		bar		"Bar"
		baz		"Baz"
		a(x)		str
		a(y)		123
		a(on)		yes
		a(off)		0
		a(subdoc)	{{"inner" : "~S:bar"}}
		a(subdoc2)	{{"inner2" : "~S:bar"}}
	}
	benchmark::test 600 200 usec our_template_vars
	benchmark::test 600 200 usec our_template_dict
	benchmark::test 600 200 usec json from_template $tmpl {
		foo		"Foo"
		bar		"Bar"
		baz		"Baz"
		a(x)		str
		a(y)		123
		a(on)		yes
		a(off)		0
		a(subdoc)	{{"inner" : "~S:bar"}}
		a(subdoc2)	{{"inner2" : "~S:bar"}}
	}
	benchmark::test 600 200 usec naive_template
	benchmark::test 300 100 usec old_template
	benchmark::test 300 100 usec new_fmt
	benchmark::test 600 200 usec our_serialize_template

	benchmark::test 600 200 usec json exists $json baz 2
	benchmark::test 600 200 usec json get $json baz 2
	benchmark::test 600 200 usec json get_typed $json baz 2
	benchmark::test 600 200 usec json extract $json baz 2

	set d	[json get $json]
	benchmark::test 600 200 usec apply {d {lindex [dict get $d baz] 2}} $d
	benchmark::test 600 200 usec apply {json {json get $json baz 2}} $json

	benchmark::test 600 200 usec nop
}

main
